aesd-char-driver/main.c

```diff
@@ -93,100 +94,152 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
-        goto out;
+        return -EFAULT;
     }
 
     newline_ptr = memchr(new_buff, '\n', count);
 
+    if (mutex_lock_interruptible(&dev->lock)) {
+        kfree(new_buff);
+        return -ERESTARTSYS;
+    }
+
     if (!newline_ptr) {
-        old_buff = (char *)dev->working_entry.buffptr;
         new_size = dev->working_entry.size + count;
-
-        dev->working_entry.buffptr = krealloc(old_buff, new_size, GFP_KERNEL);
+        dev->working_entry.buffptr = krealloc(dev->working_entry.buffptr, new_size, GFP_KERNEL);
         if (!dev->working_entry.buffptr) {
-            kfree(new_buff);
             retval = -ENOMEM;
-            goto out;
+            goto out_unlock;
         }
 
-        memcpy((void *)(dev->working_entry.buffptr + dev->working_entry.size),
-               new_buff, count);
-
+        memcpy(dev->working_entry.buffptr + dev->working_entry.size, new_buff, count);
         dev->working_entry.size = new_size;
-        retval = count;
-        kfree(new_buff);
-        goto out;
+        goto out_unlock;
     }
 
-    /* If newline found — complete command entry */
     write_size = (newline_ptr - new_buff) + 1;
     new_size = dev->working_entry.size + write_size;
 
-    dev->working_entry.buffptr =
-        krealloc((void *)dev->working_entry.buffptr, new_size, GFP_KERNEL);
+    dev->working_entry.buffptr = krealloc(dev->working_entry.buffptr, new_size, GFP_KERNEL);
     if (!dev->working_entry.buffptr) {
-        kfree(new_buff);
         retval = -ENOMEM;
-        goto out;
+        goto out_unlock;
     }
 
-    memcpy((void *)(dev->working_entry.buffptr + dev->working_entry.size),
-           new_buff, write_size);
+    memcpy(dev->working_entry.buffptr + dev->working_entry.size, new_buff, write_size);
     dev->working_entry.size = new_size;
 
     entry.buffptr = dev->working_entry.buffptr;
     entry.size = dev->working_entry.size;
-
     dev->working_entry.buffptr = NULL;
     dev->working_entry.size = 0;
 
-    /* Before adding, free the oldest entry if the buffer is full */
     if (dev->buffer.full) {
-        struct aesd_buffer_entry *oldest =
-            &dev->buffer.entry[dev->buffer.out_offs];
+        struct aesd_buffer_entry *oldest = &dev->buffer.entry[dev->buffer.out_offs];
         if (oldest->buffptr)
             kfree(oldest->buffptr);
     }
 
     aesd_circular_buffer_add_entry(&dev->buffer, &entry);
 
-    retval = count;
+out_unlock:
+    mutex_unlock(&dev->lock);
     kfree(new_buff);
+    return retval;
+}
+
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    struct aesd_dev *dev = filp->private_data;
+    loff_t newpos;
+    size_t total_size = 0;
+    uint8_t idx;
+    struct aesd_buffer_entry *entry;
+
+    if (mutex_lock_interruptible(&dev->lock))
+        return -ERESTARTSYS;
+
+    AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->buffer, idx) {
+        total_size += entry->size;
+    }
+
+    newpos = fixed_size_llseek(filp, off, whence, total_size);
```

Comment:
Shouldn't need to hold the lock on the call to fixed_size_llseek here, ideally would unlock before this.